Hide exams from guests

jonboiser opened the pull request:
## Summary
Fixes #1338 
1. If a user views `/learn/` as a guest, then there is no 'Exam' tab on top of screen
1. If a user goes directly to `learn/<channel_id>/exams` as a guest, then there is a message (and no red error box).
1. Conversely, if a user is logged in, then everything works as usual.

When writing tests, it seems that 'kolibri.lib.vue/vuex/seededshuffle' are not aliased correctly in test environment. To get around this, I manually added some files were they were assumed to be. I could try to fix this in this PR, or leave it for later.


<img width="891" alt="screen shot 2017-04-26 at 2 34 53 pm" src="https://cloud.githubusercontent.com/assets/10248067/25450709/c5c3fe54-2a8d-11e7-9dfc-24e61631e3cd.png">
<img width="899" alt="screen shot 2017-04-26 at 2 34 46 pm" src="https://cloud.githubusercontent.com/assets/10248067/25450708/c5c3c434-2a8d-11e7-9b1c-d5d65d1a30be.png">

rtibbles commented:
@jonboiser I have a fix for the incorrect aliasing in the test environment in this PR: https://github.com/learningequality/kolibri/pull/1324 it's a low risk merge, because it just adds additional tests, so if you want to give a quick review and merge there we can resolve it.

rtibbles commented on kolibri/core/assets/src/core-app/seededshuffle.js:
```diff
@@ -0,0 +1 @@
+module.exports = require('seededshuffle');
```
Maybe lets get https://github.com/learningequality/kolibri/pull/1324 merged, and remove these sooner rather than later!

rtibbles commented on kolibri/plugins/learn/assets/src/state/actions.js:
```diff
@@ -447,10 +447,17 @@ function showSearch(store, channelId, searchTerm) {
-  coreActions.setChannelInfo(store, channelId).then(
+  // if user is not logged in, this action is a noop
+  if (!userIsLoggedIn) {
+    store.dispatch('CORE_SET_PAGE_LOADING', false);
```
Could we maybe add an error message here too, just to give the user some indication that something has gone wrong? Either that, or just redirect back to base learn.

jonboiser replied:
Can we do the redirect inside the router, before we even hit this action? This would obviate a lot of this code actually, although it's debatable whether it is good to just redirect a user without explanation.

I could dispatch an error message to show in the red box, but it seems like the "did you forget to login" message should be enough explanation.

indirectlylit replied:
This is in a category of route/action handling that would be great to approach more systematically than we currently are. At a high level, I think we have a need for middleware wrappers around routing actions to handle cases like:

* 'not found'-type issues (e.g. content URLs, reports for invalid IDs, etc)
* authorization issues (e.g. loading a manage page as a regular user)
* general uncaught errors (current fallback on `SET_ERROR`)

A lot of these are kind of the client-side version of what you'd expect to see from a properly configured web server (404s, 403s, 500s...) and we might be able to think about handling them similarly (e.g. adding "decorators" to various actions specifying what kind of user it's applicable for)

In general, I'd like to avoid redirecting for any of these because it dissociates the URL that caused the issue with the result. It means if someone wants to e.g. refresh the page to try again, it won't work. Instead, we should move towards incorporating common scenarios into `core-base` and handling the scenarios consistently across routing actions.

Redirecting might sometimes be appropriate - like directly after logout - but I think the default should be displaying a message without changing the URL.

indirectlylit replied:
Of course - all that is forward looking :) for the current issue, I'm sure there's an uglier but quicker fix and redirecting wouldn't be the end of the world!

rtibbles commented on kolibri/plugins/learn/assets/src/views/exam-list/index.vue:
```diff
@@ -1,26 +1,34 @@
-        <p class="answer-count">{{ $tr('howManyCorrect', { score: exam.score, outOf: exam.questionCount })}}</p>
-        <div class="button-or-score">
-          <b>{{ $tr('percentCorrect', { pct: exam.score/exam.questionCount })}}</b>
+    <!-- TODO this HTML is duplicated in a number of places, maybe make a component? -->
```
Can you open an issue for this and remove the TODO?

jonboiser commented:
After review:

1. Deleted shims for improperly-aliased node modules
1. Removed TODO and added an issue to componentize the 'did you forget to logout?' screen
1. Left the 'showExamListPage' action as is (as noop if user is not authenticated).

indirectlylit commented:
>  added an issue to componentize the 'did you forget to logout?' screen

ref: #1365
